pyanaconda/installation_tasks.py

```diff
@@ -148,187 +119,93 @@ def status_message(self):
-            self.task_started.disconnect(self._parent.task_started.emit)
-            self.task_completed.disconnect(self._parent.task_completed.emit)
-        # set the parent
-        self._parent = parent_item
-        # Connect own signals "up" to the parent,
-        # so that it is possible to monitor how all nested TaskQueues and Tasks
-        # are running from the top-level element.
-
-        # connect own start/completion signal to parents queue start/completion signal
-        self.started.connect(self._parent.queue_started.emit)
-        self.completed.connect(self._parent.queue_completed.emit)
-
-        # propagate start/completion signals from nested queues/tasks
-        self.queue_started.connect(self._parent.queue_started.emit)
-        self.queue_completed.connect(self._parent.queue_completed.emit)
-        self.task_started.connect(self._parent.task_started.emit)
-        self.task_completed.connect(self._parent.task_completed.emit)
-
-    def start(self):
-        """Start processing of the task queue."""
-        do_start = False
-        with self._lock:
-            # the task queue can only be started once
-            if self.running or self.done:
-                if self.running:
-                    # attempt to start a task that is already running
-                    log.error("Can't start task queue %s - already running.")
-                else:
-                    # attempt to start a task that an already finished task
-                    log.error("Can't start task queue %s - already done.")
-            else:
-                do_start = True
-                self._running = True
-                if self.task_count:
-                    # only set the initial task number if we have some tasks
-                    self._current_task_number = 0
-                    self._current_queue_number = 0
-                else:
-                    log.warning("Attempting to start an empty task queue (%s).", self.name)
-
-        if do_start:
-            # go over all task groups and their tasks in order
-            self.started.emit(self)
-            if len(self) == 0:
-                log.warning("The task group %s is empty.", self.name)
-            for item in self:
-                # start the item (TaskQueue/Task)
-                item.start()
-
-            # we are done, set the task queue state accordingly
-            with self._lock:
-                self._running = False
-                self._done = True
-                # also set the current task variables accordingly as we no longer process a task
-                self._current_task_number = None
-                self._current_queue_number = None
-
-            # trigger the "completed" signals
-            self.completed.emit(self)
+    def _run(self):
+        """Run the task queue."""
+        for item in self._queue:
+            # start the item (TaskQueue/Task)
+            item.start()
```

Comment:
You start the item and connect the signals only later, I think? I don't see why this can work?

Reply:
It is a little messy, the signals are connected when you append an item to a queue.